test/e2e/upgrades/apps/replicasets.go

```diff
@@ -0,0 +1,145 @@
+	ns := f.Namespace.Name
+	nginxImage := imageutils.GetE2EImage(imageutils.NginxSlim)
+
+	By(fmt.Sprintf("Creating a replicaset %q in namespace %q", rsName, ns))
+	replicaSet := framework.NewReplicaSet(rsName, ns, int32(1), map[string]string{"test": "upgrade"}, "nginx", nginxImage)
+	rs, err := c.Extensions().ReplicaSets(ns).Create(replicaSet)
+	framework.ExpectNoError(err)
+
+	By(fmt.Sprintf("Waiting replicaset %q's container image to be updated", rsName))
+	err = framework.WaitForReplicaSetImage(c, ns, rsName, nginxImage, interval, timeout)
+	framework.ExpectNoError(err)
+
+	By(fmt.Sprintf("Waiting for replicaset %q to have all of its replicas ready.", rsName))
+	framework.ExpectNoError(framework.WaitForReadyReplicaSet(c, ns, rsName))
+
+	// Store the old RS status.
+	r.oldRSStatus = rs.Status
+	oldUID := rs.UID
+
+	// Trigger a new rollout so that we have some history.
+	By(fmt.Sprintf("Triggering a new rollout for replicaset %q", rsName))
+	rs, err = framework.UpdateReplicaSetWithRetries(c, ns, rsName, func(update *extensions.ReplicaSet) {
+		update.Spec.Template.Spec.Containers[0].Name = "updated-nginx"
+	})
+	framework.ExpectNoError(err)
+
+	// Use observedGeneration to determine if the controller noticed the pod template update.
+	framework.Logf("Wait replicaset %q to be observed by the replicaset controller", rsName)
+	framework.ExpectNoError(framework.WaitForObservedReplicaSet(c, ns, rsName, rs.Generation, interval, timeout))
+
+	By(fmt.Sprintf("Waiting for replicaset %q to have all of its replicas ready.", rsName))
+	framework.ExpectNoError(framework.WaitForReadyReplicaSet(c, ns, rsName))
+
+	rs, err = c.Extensions().ReplicaSets(ns).Get(rs.Name, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	if rs == nil {
+		framework.ExpectNoError(fmt.Errorf("expected a replicaset"))
+	}
+
+	// Store new RS status.
+	r.newRSStatus = rs.Status
+
+	if rs.UID != oldUID {
+		framework.ExpectNoError(fmt.Errorf("expected getting the same replicaset after the new rollout"))
+	}
+}
+
+// Test checks whether the replicasets are the same after an upgrade.
+func (r *ReplicaSetUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	c := f.ClientSet
+	ns := f.Namespace.Name
+
+	// Block until upgrade is done
+	By(fmt.Sprintf("Waiting for upgrade to finish before checking replicaset %q", rsName))
+	<-done
+
+	By(fmt.Sprintf("Getting new status for replicaset %q after upgrade is done", rsName))
+	newRS, err := c.Extensions().ReplicaSets(ns).Get(rsName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	if newRS == nil {
+		framework.ExpectNoError(fmt.Errorf("expected a replicaset %q, but got nil", rsName))
+	}
+	newRSStatus := newRS.Status
+
+	By(fmt.Sprintf("Waiting for replicaset %q to have all of its replicas ready.", rsName))
+	framework.ExpectNoError(framework.WaitForReadyReplicaSet(c, ns, rsName))
+
+	By(fmt.Sprintf("Checking that new status of replicaset %q is the same as after the upgrade", rsName))
```

Comment:
I'm not sure it makes sense to expect the entire Status to be the same before and after the upgrade.

Other than making sure the Pods are Ready (as you did above), let's just check:

* The UID is the same (it's the same RS).
* We can scale it up (by 1 is fine) and the new Pods become Ready too (the controller is active).

Reply:
+1 for UID similarity.
Never thought about scaling - will be a good metric to add. Thanks!